### PR TITLE
Implement the column collation for Mysql

### PR DESCRIPTION
**Pull request by mikeSimonson**
Related to #2538

### CODE REVIEW

**Comment by Ocramius** [comment]
@mikeSimonson should this fix (since it fixes #2538, if I understand it correctly) come with a test case then?

**Comment by tristanbes** [comment]
Any updates on this @mikeSimonson ? I can confirm that this fix should fix my issue about Doctrine Migrations generating a diff without the collate information set on `Table` option

` * @ORM\Table(name="fos_user", options={"collate"="utf8mb4_unicode_ci", "charset"="utf8mb4"}) `

When I put a new field and I generate the diff it outputs
`$this->addSql('ALTER TABLE fos_user ADD lessNoise TINYINT(1) NOT NULL');`

resulting in the field created in `utf8_unicode` and `utf8_unicode_ci` for collation instead of the settings set on the table.

**Comment by mikeSimonson** [comment]
@Ocramius Do you mean that the test should be removed or that it needs a higher level test case ?

**Comment by Ocramius** [comment]
@mikeSimonson higher level test case reflecting #2538

**Comment by mikeSimonson** [comment]
@Ocramius What is the right approach here for the test ?

File: tests/Doctrine/Tests/DBAL/Schema/SchemaTest.php [hunk 1]
```diff
@@ -449,4 +451,48 @@ public function testVisitsNamespaceVisitor()
...
+            ['CREATE TABLE public.foo (test VARCHAR(255) NOT NULL COLLATE utf8_unicode_ci) DEFAULT CHARACTER SET utf8 COLLATE utf8_unicode_ci ENGINE = InnoDB'],
+            $schema->toSql($platform),
+            $platform->getColumnDeclarationSQL('test')
+        );
+    }
+
+    /**
+     * https://github.com/doctrine/dbal/issues/2538
+     */
+    public function testFixGithub2538_autoincrement()
```
Comment (Ocramius) [inline tests/Doctrine/Tests/DBAL/Schema/SchemaTest.php]: This test seems unrelated to the changes, or am I missing something?

File: tests/Doctrine/Tests/DBAL/Schema/SchemaTest.php [hunk 2]
```diff
@@ -449,4 +451,48 @@ public function testVisitsNamespaceVisitor()
...
+    {
+        $this->markTestSkipped();
+
+        $tableName = "public.foo";
```
Comment (Ocramius) [inline tests/Doctrine/Tests/DBAL/Schema/SchemaTest.php]: Single quotes

File: tests/Doctrine/Tests/DBAL/Schema/SchemaTest.php (hunk 1 above)
Comment (Ocramius) [inline tests/Doctrine/Tests/DBAL/Schema/SchemaTest.php]: Please use a method name describing the issue being covered. Also, use `@group #2538` for labeling it

File: tests/Doctrine/Tests/DBAL/Schema/SchemaTest.php [hunk 3]
```diff
@@ -449,4 +451,48 @@ public function testVisitsNamespaceVisitor()
 
         $this->assertNull($schema->visit($visitor));
     }
+
+    /**
+     * https://github.com/doctrine/dbal/issues/2538
+     */
+    public function testFixGithub2538Collation()
```
Comment (Ocramius) [inline tests/Doctrine/Tests/DBAL/Schema/SchemaTest.php]: Please use a method name describing the issue being covered. Also, use `@group #2538` for labeling it

File: tests/Doctrine/Tests/DBAL/Platforms/AbstractPlatformTestCase.php [hunk 4]
```diff
@@ -1429,4 +1429,18 @@ public function getGeneratesFloatDeclarationSQL()
             array(array('precision' => 8, 'scale' => 2), 'DOUBLE PRECISION'),
         );
     }
+
+    public function testFixGithub2538Collation()
```
Comment (Ocramius) [inline tests/Doctrine/Tests/DBAL/Platforms/AbstractPlatformTestCase.php]: Please use a method name describing the issue being covered. Also, use `@group #2538` for labeling it

**Review by Ocramius — CHANGES_REQUESTED** [review]
One last change probably needed: verifying that the comparator will pick up the collation (unless that is already covered)

**Comment by yukoff** [comment]
What is the progress with this one? And maybe there should be column charset implemented ass well?

**Comment by Tobion** [comment]
I don't see the point of this PR. Column collations are working in mysql already in general.

`@ORM\Column(type="string", options={"collation"="ascii_general_ci"})` will correctly generate a mysql column like `myfield VARCHAR(255) NOT NULL COLLATE ascii_general_ci`

The requested change seems useless. MysqlPlatform already implements 

```
public function supportsColumnCollation()
{
        return true;
}
```
and the rest is handled by the default implementation of AbstractPlatform.

There are two bugs with column collations:
1. schema update / migrations doesn't recognize collation changes, see #2401 similar issues
2. When you have a column collation overwritten for an enity primary key that is used as forgeign key in another entity, then the collation is MISSING in the foreign key definition. This makes the generated schema invalid because the collations can be different which is not allowed in foreign key constraints. This is what is stumbled upon. I'll create an issue for this.

**Comment by Ocramius** [comment]
> Column collations are working in mysql already in general.

Tests seem to say otherwise, but feel free to provide more tests or comment on the ones provided in this PR.

**Pull request closed without merging**